graphql_compiler/compiler/emit_sql.py

```diff
@@ -239,19 +340,152 @@ def start_global_operations(self):
+        self._fold_outer_vertex_location = self._current_location
+        self._fold_inner_vertex_location = fold_scope_location
+
+        # basic info about the outer vertex and inner vertex
+        inner_vertex_table = self._sql_schema_info.vertex_name_to_table[
+            self._ir.query_metadata_table.get_location_info(fold_scope_location).type.name
+        ].alias()
+        outer_vertex_table = self._current_alias.alias()
+        inner_vertex_type = self._ir.query_metadata_table \
+            .get_location_info(fold_scope_location) \
+            .type \
+            .name
+        outer_vertex_type = self._current_classname
+
+        self._aliases[
+            (self._current_location.query_path, self._fold_inner_vertex_location)
+        ] = inner_vertex_table
+
+        # collect edge information to join the fold subquery to the main selectable
+        edge_direction = fold_scope_location.fold_path[0][0]
+        edge_name = fold_scope_location.fold_path[0][1]
+        full_edge_name = '{}_{}'.format(edge_direction, edge_name)
+
+        edge = self._sql_schema_info.join_descriptors[
+            self._current_classname][full_edge_name]
+
+        # we will use this to make sure that the join inside
+        # the subquery uses one alias of the current table
+        # and that the join outside the subquery uses another alias
+        join_info = (edge,
+                     outer_vertex_table,
+                     inner_vertex_type,
+                     self._current_alias)
+
+        self._fold = SQLFoldObject(inner_vertex_table,
+                                   outer_vertex_table,
+                                   inner_vertex_type,
+                                   outer_vertex_type,
+                                   join_info)
+        # find outputs for this fold in self._all_folded_outputs and add to self._fold_outputs
+        if fold_scope_location.fold_path in self._all_folded_outputs:
+            for fold_output in self._all_folded_outputs[fold_scope_location.fold_path]:
+                if fold_output.field == '_x_count':
+                    raise NotImplementedError(u'_x_count not implemented in SQL')
+
+                intermediate_fold_output_name = self._alias_generator.gen_output_column()
+                self._fold_output_column_aliases[fold_output] = intermediate_fold_output_name
+                # add array aggregated intermediate output to self._fold_outputs
+                self._fold.outputs.append(
+                    sqlalchemy.func.array_agg(
+                        self._fold.inner_vertex_table.c[fold_output.field]
+                    ).label(intermediate_fold_output_name)
+                )
+
+        # the group by field is the unique identifier for the outer vertex
+        self._fold.outputs.append(self._fold.outer_vertex_table.c[edge.from_column])
+        self._fold.group_by.append(self._fold.outer_vertex_table.c[edge.from_column])
```

Comment:
This is a [code smell](https://en.wikipedia.org/wiki/Code_smell). Just because some state got encapsulated into `self._fold` of type `SQLFoldObject`, it doesn't mean that this is composable or understandable code. And yes, you may have avoided the linter's complaint about mutating object internal state by making it public, but that is treating the symptoms and not the underlying problem.

Here's the underlying problem: `def fold()` is almost a 100 lines long, and is full of extremely complex code that I'm finding quite difficult to understand already. Code only becomes less understandable over time, and that means trouble.

Instead, try to describe in a few words what this function needs to do. You'll notice that you'll need to define some terminology to keep the description concise — great! Make helper functions for those bits of terminology, and go from there.

Reply:
I agree - I think the most recent PR addresses most of these issues. It seems like we would prefer the SQLFoldObject to be built to completion all at once, as opposed to built incrementally. I think that is possible for folds with no inner traversals, but won't generalize because different pieces of information will be stored in different parts of the IR, e.g. each traversal within a fold adds another join descriptor and table to our SQLFoldObject. Hopefully the current design represents a good middle ground. I don't think it will need substantial changes to handle the general, multi traversal fold case, but it looks as build-to-completion as it can.